### ISSUE TITLE
Remove non-printing Unicode characters from names

### ISSUE THREAD

**Issue by jonkerz** [calacademy-research/antcat#168]
Split out from https://github.com/calacademy-research/antcat/issues/128#issuecomment-217120654

This whole issue looks very much like source code, because it is. To be finished in the future. I'm pasting it here as I've already procrastinated it since May.

```ruby
###############################################################
#
# See https://github.com/calacademy-research/antcat/issues/128#issuecomment-217120654
# Basically, there's Unicode crap in four names, and
# it's not possible to remove then via the GUI.
#
# Search for 'beginning with' "carinata" and there's no "Carebara carinata"
# in the results, because there's a Unicode character before "carinata":
# http://antcat.org/catalog/search/quick_search?utf8=%E2%9C%93&search_type=beginning_with&qq=carinata&valid_only=true
#
# On the other hand, searching for species 'containing' "carinata" works:
# http://antcat.org/catalog/search/quick_search?utf8=%E2%9C%93&search_type=containing&qq=carinata&valid_only=true
#
# See it in action in the Rails console:
# puts Taxon.find(461558).name_html_cache
#
###############################################################
#
# Step 1: Find tainted names
#
# Name.find_each do |name|
#   puts "Name.find(#{name.id})" if name.name =~ /[^[:print:]]/
# end
#
# Name.find(168950)
# Name.find(168952) # <-- SubspeciesName / Protonym
# Name.find(168954)
# Name.find(168955)
#
# Three of them are SpeciesNames linked to a Species.
# Taxon.find([461558, 461562, 461563])
#
# One is a SubspeciesName linked to a Protonym.
# Protonym.find(198466).
# Buuuut, the name is really a species name: http://antcat.org/catalog/461561?qq=Carebara+hornata and it was described as such in Bharti & Kumar (2013), http://antcat.org/references/142470
# This should probably be fixed in another issue.
#
###############################################################
#
# Step 2: The migration
#

class RemoveUnicodeCharacters < ActiveRecord::Migration
  def up
    # Manually update all fields and caches
    Taxon.find([461558, 461562, 461563]).each do |taxon|
      tn = taxon.name
      tn.name = remove_u0014 tn.name
      tn.name_html = remove_u0014 tn.name_html
      tn.epithet = remove_u0014 tn.epithet
      tn.epithet_html = remove_u0014 tn.epithet_html
      tn.protonym_html = remove_u0014 tn.protonym_html
      tn.save!
    end

    # Special case as `Protonym` names are not cached, and special case
    # again as the really-a-species-name is stored as a `SubspeciesName`,
    # and another minor special case as the Unicode character is trailing.
    # hornata = SubspeciesName.find(168952)
    # TODO.........
  end

  def down
    # Do nothing. If we *ever* need to restore the Unicode
    # crap, let's figure it out at that time.
  end

  private
    def remove_u0014 string
      string.delete "\u0014"
    end
end
```

**Comment by jonkerz** [comment]
Long story short: This code fixes the first three names, but then I discovered a fourth name in a protonym, and that name is stored as a subspecies name, but it should probably be a species name, and that may be an ever biggier issue. Here's the fourth name btw: http://antcat.org/catalog/461561?qq=Carebara+hornata

The best way to proceed is probably to ignore the last name and fix the rest. To be continued.

**Comment by jonkerz** [comment]
Implemented in 3b2ea623059b0628a82f273bc6eae7e30aa0d040 using a different method.
